Web deploy

aleksandra-tarkowska opened the pull request:
This PR improve web plugin starting and stopping OMERO.web app

To test:
- start web twice, you should get error message saying already started.
- check if ``WARNINGS: ?: (1_8.W001) The standalone TEMPLATE_* settings...`` disappeared. When deploying with development server you will see ``System check identified no issues (1 silenced).``
- check if ``dist/var/django_secret_key`` file has correct permission (CHECK WINDOWS!):

    ```
-rw-------   1 omero  staff   50 24 Oct 14:58 django_secret_key
```
- run ``bin\omero web iis -h`` and check if you see only:

 ```
 usage: dist/bin/omero web iis [-h] [--remove] [--keep-sessions | --no-wait]
 ...
 ```
- run ``bin\omero web start|restart -h`` and check if you get 

 ```
 usage: dist/bin/omero web start [-h] [--keep-sessions | --no-wait]
                                [--workers WORKERS]
                                [--worker-connections WORKER_CONNECTIONS]
                                [--wsgi-args WSGI_ARGS]
 ...
  --workers WORKERS                     NGINX only: the number of worker processes for handling requests.
   --worker-connections WORKER_CONNECTIONS
                                        NGINX only: the maximum number of simultaneous clients.
   --wsgi-args WSGI_ARGS                 NGINX only: additional arguments. Check Gunicorn Documentationhttp://docs.gunicorn.org/en/latest/settings.html
 ```

joshmoore commented:
```
Clearing expired sessions. This may take some time...
Traceback (most recent call last):
  File "manage.py", line 56, in <module>
    execute_from_command_line(sys.argv)
  File "/opt/v/lib/python2.7/site-packages/django/core/management/__init__.py", line 351, in execute_from_command_line
    utility.execute()
  File "/opt/v/lib/python2.7/site-packages/django/core/management/__init__.py", line 343, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/opt/v/lib/python2.7/site-packages/django/core/management/base.py", line 394, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/opt/v/lib/python2.7/site-packages/django/core/management/base.py", line 444, in execute
    self.check()
  File "/opt/v/lib/python2.7/site-packages/django/core/management/base.py", line 491, in check
    warnings = [e for e in all_issues if checks.WARNING <= e.level < checks.ERROR and not e.is_silenced()]
  File "/opt/v/lib/python2.7/site-packages/django/core/checks/messages.py", line 58, in is_silenced
    return self.id in settings.SILENCED_SYSTEM_CHECKS
TypeError: argument of type 'bool' is not iterable
```

aleksandra-tarkowska commented:
@joshmoore how did you get that error? + web dependencies?

aleksandra-tarkowska commented:
@joshmoore sorry that was miss usage https://docs.djangoproject.com/en/1.8/ref/settings/#silenced-system-checks. Now should be better

sbesson commented:
With #4290 and #4295 now merged, we can likely remove the `--breaking` comment on this PR and include it into today's merge build if it's ready.

aleksandra-tarkowska commented:
Lets try when travis is green

aleksandra-tarkowska commented:
succesfully deployed using https://ci.openmicroscopy.org/view/DEV/job/OMERO-DEV-merge-deploy/137/console

aleksandra-tarkowska commented:
Warning disappeard https://ci.openmicroscopy.org/view/DEV/job/OMERO-DEV-merge-integration-Python27/25/consoleFull 
```
12:14:17 System check identified some issues:
12:14:17 
12:14:17 WARNINGS:
12:14:17 ?: (1_8.W001) The standalone TEMPLATE_* settings were deprecated in Django 1.8 and the TEMPLATES dictionary takes precedence. You must put the values of the following settings into your default TEMPLATES dict: TEMPLATE_DIRS, TEMPLATE_CONTEXT_PROCESSORS.
12:14:17 Clearing expired sessions. This may take some time... 
12:14:17 [OK]
12:14:17 Starting OMERO.web... [OK]
```
vs https://ci.openmicroscopy.org/view/DEV/job/OMERO-DEV-merge-integration-Python27/27/consoleFull

```
15:28:41 Clearing expired sessions. This may take some time... 
15:28:41 [OK]
15:28:41 Starting OMERO.web... [OK]
```

aleksandra-tarkowska commented:
description was updated

joshmoore commented:
@aleksandra-tarkowska : how many more commits are planned? (So I know when to start testing)

aleksandra-tarkowska commented:
@joshmoore Sorry I didn;t know you were planing to test it over the weekend ;-) I am done now.

joshmoore commented:
* secret key is read only :thumbsup:
 * iis help is improved with start still working :thumbsup:
 * warning seems to be gone. : thumbsup:
 * output of restart  is **non-deterministic**. :thumbsdown: Running it several times after one another, I get both:
   * `/opt/webspace/OMERO.server-5.2.0-rc1-27-bf383cb-ice35-b137/var/django.pid exists (PID: 25753)! Use 'web stop or restart' first.` and
   * `Removed stale /opt/webspace/OMERO.server-5.2.0-rc1-27-bf383cb-ice35-b137/var/django.pid`

If fixing the latter is an issue for RC2, we might want to cherry pick from this PR to keep things simple.

aleksandra-tarkowska commented:
Last commit can be tested:

```
$ dist/bin/omero web start
Post-processed 'omeroweb.viewer.min.css' as 'omeroweb.viewer.min.css'
Post-processed 'omeroweb.viewer.min.js' as 'omeroweb.viewer.min.js'

0 static files copied to '/Users/ola/OMERO/openmicroscopy/dist/lib/python/omeroweb/static', 618 unmodified, 2 post-processed.
Clearing expired sessions. This may take some time... [OK]
Starting OMERO.web... [OK]
```

Open localhost:4080 (you want see statics)

```
$ dist/bin/omero web status
OMERO.web status... [RUNNING] (PID 7272)

$ cat dist/var/django.pid 
7272

$ dist/bin/omero web start
Post-processed 'omeroweb.viewer.min.css' as 'omeroweb.viewer.min.css'
Post-processed 'omeroweb.viewer.min.js' as 'omeroweb.viewer.min.js'

0 static files copied to '/Users/ola/OMERO/openmicroscopy/dist/lib/python/omeroweb/static', 618 unmodified, 2 post-processed.
Clearing expired sessions. This may take some time... [OK]
Starting OMERO.web... [FAILED] OMERO.web already started. /Users/ola/OMERO/openmicroscopy/dist/var/django.pid exists (PID: 7272)! Use 'web stop or restart' first.

$ dist/bin/omero web stop
Stopping OMERO.web... [OK]
OMERO.web WSGI workers (PID 7272) killed.
Removed stale /Users/ola/OMERO/openmicroscopy/dist/var/django.pid

$ dist/bin/omero web status
OMERO.web status... [NOT STARTED]

$ dist/bin/omero web stop
Stopping OMERO.web... [NOT STARTED]
```

test bad behavior, wait few sec between steps to make s

```
$ dist/bin/omero web start
Post-processed 'omeroweb.viewer.min.css' as 'omeroweb.viewer.min.css'
Post-processed 'omeroweb.viewer.min.js' as 'omeroweb.viewer.min.js'

0 static files copied to '/Users/ola/OMERO/openmicroscopy/dist/lib/python/omeroweb/static', 618 unmodified, 2 post-processed.
Clearing expired sessions. This may take some time... [OK]
Starting OMERO.web... [OK]
```

LOG IN TO WEB localhost:4080 (you want see statics)

```
$ dist/bin/omero web status
OMERO.web status... [RUNNING] (PID 7439)

$ cat dist/var/django.pid 
7439

$ rm -f dist/var/django.pid

$ dist/bin/omero web start
Post-processed 'omeroweb.viewer.min.css' as 'omeroweb.viewer.min.css'
Post-processed 'omeroweb.viewer.min.js' as 'omeroweb.viewer.min.js'

0 static files copied to '/Users/ola/OMERO/openmicroscopy/dist/lib/python/omeroweb/static', 618 unmodified, 2 post-processed.
Clearing expired sessions. This may take some time... [OK]
Starting OMERO.web... [OK]
```

By that time it is unknown that there are any stale processes running
REFRESH WEB

```
$ cat dist/var/django.pid 
7484

$ dist/bin/omero web status
OMERO.web status... [ERROR] OMERO.web workers (PID 7484) - no such process. Use `ps aux| grep /Users/ola/OMERO/openmicroscopy/dist/var/django.pid` and kill stale processes by hand. 
[NOT STARTED]
```

```
$ dist/bin/omero web stop
Stopping OMERO.web... [ERROR] OMERO.web workers (PID 7484) - no such process. Use `ps aux| grep /Users/ola/OMERO/openmicroscopy/dist/var/django.pid` and kill stale processes by hand. 
Removed stale /Users/ola/OMERO/openmicroscopy/dist/var/django.pid
```

Test apache

```
$ dist/bin/omero config set omero.web.application_server wsgi
$ dist/bin/omero web start
Post-processed 'omeroweb.viewer.min.css' as 'omeroweb.viewer.min.css'
Post-processed 'omeroweb.viewer.min.js' as 'omeroweb.viewer.min.js'

0 static files copied to '/Users/ola/OMERO/openmicroscopy/dist/lib/python/omeroweb/static', 618 unmodified, 2 post-processed.
Clearing expired sessions. This may take some time... [OK]
You are deploying OMERO.web using apache and mod_wsgi. Add config to Apache and reload.

$ ps aux| grep /Users/ola/OMERO/openmicroscopy/dist/var/django.pid
ola             7636   0.0  0.0  2432768    648 s000  R+   10:59am   0:00.00 grep /Users/ola/OMERO/openmicroscopy/dist/var/django.pid

$ dist/bin/omero web status
OMERO.web status... You are deploying OMERO.web using apache and mod_wsgi. Cannot check status.

$ dist/bin/omero web stop
Stopping OMERO.web... You are deploying OMERO.web using apache and mod_wsgi. Cannot check status.
```

sbesson commented:
Given the complexity of the testing scenario and the fact that all the other improvements have been signed off by @joshmoore, can we split the `start/restart/stop` improvements into a separate PR so that we can at least get the other changes in?

aleksandra-tarkowska commented:
I am sorry that will be tricky no. The last comment looks big but these are really few commands only.

aleksandra-tarkowska commented:
try this now

sbesson commented:
With the last state of this PR, the `web restart` command is broken for me locally:

```
sbesson@lifesci-37398:OMERO.server-5.2.0-rc1-30-f150c98-ice35-b138 $ bin/omero web start
Post-processed 'omeroweb.viewer.min.css' as 'omeroweb.viewer.min.css'
Post-processed 'omeroweb.viewer.min.js' as 'omeroweb.viewer.min.js'

0 static files copied to '/Users/sbesson/Downloads/OMERO.server-5.2.0-rc1-30-f150c98-ice35-b138/lib/python/omeroweb/static', 618 unmodified, 2 post-processed.
Clearing expired sessions. This may take some time... [OK]
Starting OMERO.web... [OK]
sbesson@lifesci-37398:OMERO.server-5.2.0-rc1-30-f150c98-ice35-b138 $ bin/omero web restart
Stopping OMERO.web... [OK]
OMERO.web WSGI workers (PID 22978) killed.
Removed stale /Users/sbesson/Downloads/OMERO.server-5.2.0-rc1-30-f150c98-ice35-b138/var/django.pid
sbesson@lifesci-37398:OMERO.server-5.2.0-rc1-30-f150c98-ice35-b138 $ 
````

sbesson commented:
Retested using the latest version of the build. All three first comments from https://github.com/openmicroscopy/openmicroscopy/pull/4294#issuecomment-151403515 still apply. Ready to merge for 5.2.0-rc1 @jburel

jburel commented:
Thanks all
